### ISSUE TITLE
Question: ec goal conversion - chunkserver requirements

### ISSUE THREAD

**Issue by eleaner** [lizardfs/lizardfs#767]
Hi Development team.

Could someone please clarify the requirements for ec goal conversion?
The scenario:
4 chunkservers, ec2,2 
I want to add 5th chunkserver and move to ec3,2

is it enough to add one chunkserver, change the goal and wait?
Or, do I need to add five chunkservers so the ec2,2 can effectively be replicated to another 3,2 copy before it gets deleted. Then only after conversion, I would remove four chunkservers I used previously.

Which one is it?

**Comment by asyslinux** [comment]
I have a similair question, but have a different configuration:

I have over 120 million small files in this setup:

10 10 : $ec(4,2) { chunk1 chunk2 chunk3 chunk4 chunk5 chunk6 }

chunk1 = server-1 / mfschunkserver-1
chunk2 = server-2 / mfschunkserver-2
chunk3 = server-3 / mfschunkserver-3
chunk4 = server-4 / mfschunkserver-4
chunk5 = server-5 / mfschunkserver-5
chunk6 = server-6 / mfschunkserver-6

If i add new server to this configuration:

1. Add new 7th server with name chunk1 as mfschunkserver. But I think with this variant I can not expand effective space entire cluster, and need install 6 new servers for entirely expand cluster. 

Or

2. Add new server with name chunk7 as mfschunkserver and expand mfsgoal.cfg to this configuration:

10 10 : $ec(4,2) { chunk1 chunk2 chunk3 chunk4 chunk5 chunk6 chunk7 }

Need to change parity to $ec(5,2) or I can leave $ec(4,2) with 7 servers? 
Change goal and restart mfsmaster`s ? dangerous?

**Comment by jkiebzak** [comment]
@eleaner - I believe if you want to move from ec2,2 to ec3,2, you will need add five new servers (or at least 5 new daemons). 

@asyslinux - if you simply want to "expand effective space entire cluster", you can added another chunkserver (aka chunk7). There is no need to change the erasure coding. If you use ec4,2 with 7 (or more) chunkservers, the 6 pieces (4+2) of the chunk will be written to 6 different chunk servers. When adding a new chunkserver to the existing cluster, the chunkservers will start to rebalance - move existing chunks to the new server until all servers are equally used.

**Comment by asyslinux** [comment]
> @eleaner - I believe if you want to move from ec2,2 to ec3,2, you will need add five new servers (or at least 5 new daemons).
> 
> @asyslinux - if you simply want to "expand effective space entire cluster", you can added another chunkserver (aka chunk7). There is no need to change the erasure coding. If you use ec4,2 with 7 (or more) chunkservers, the 6 pieces (4+2) of the chunk will be written to 6 different chunk servers. When adding a new chunkserver to the existing cluster, the chunkservers will start to rebalance - move existing chunks to the new server until all servers are equally used.

But in my configuration i have a hardcoded names of chunkservers:
{ chunk1 chunk2 chunk3 chunk4 chunk5 chunk6 }

What I need do in this case?

**Comment by jkiebzak** [comment]
do you need the hard coded names? Does it server some purpose? If not, I would revert the configuration to `10 10 : $ec(4,2)` and reload the master (I can't remember if `reload` is good enough to reload the goals... it may require a `restart`).

**Comment by asyslinux** [comment]
Yes I need, because first time I used 2 physical servers with 6 mfschunkservers with 4 hdd`s per 1 chunkserver, and for properly balancing data in this case i need to set hardcoded names. But later I move 4 chunkservers with data on new 4 servers, but save old configuration with hardcoded names of chunkservers.

Are you sure with revert goal configuration to variant without names and this is change not destroy my data? May be safest way - make new goal 15 15 : $ec(4,2) without names, then run mfssetgoal -r 15 /my/mounted/lizardfs ?

**Comment by jkiebzak** [comment]
i have removed the label requirement from goals in the past without any issues.

I believe running `mfssetgoal -r 15 /my/mounted/lizardfs` will force a rewrite of all the chunks (aka, rewrite all your data)

**Comment by asyslinux** [comment]
Ok, thanks for reply, first I try to test this on test environment, need to deploy and check.

**Comment by asyslinux** [comment]
> i have removed the label requirement from goals in the past without any issues.
> 
> I believe running `mfssetgoal -r 15 /my/mounted/lizardfs` will force a rewrite of all the chunks (aka, rewrite all your data)

I made test already:

if I change goal `10 10 : $ec(4,2) { chunk1 chunk2 chunk3 chunk4 chunk5 chunk6 }`
to `15 15 : $ec(4,2)` with mfssetgoal -r 15 /my/mounted/lizardfs , they do not rewrite data, its smart if goal same same.

if I change goal only in mfsgoals.cfg from  `10 10 : $ec(4,2) { chunk1 chunk2 chunk3 chunk4 chunk5 chunk6 }` to  `10 10 : $ec(4,2)` , and reload mfsmaster`s, restart not needed, then no have any problems.

All variants is acceptable in this case.

**Comment by jkiebzak** [comment]
Thanks for testing 



null
